.github/workflows/canister-tests.yml

```diff
@@ -195,26 +195,120 @@ jobs:
-          # stuff but the cache is anyway invalidated on every toolchain update.
-          # We also include all Cargo tomls and the lockfile
-          path: |
-            ~/.cargo
-            target
-          key: ${{ runner.os }}-cargo-${{ hashFiles('Cargo.lock', 'Cargo.toml', 'src/**/Cargo.toml', 'rust-toolchain.toml') }}-3
+          path: /tmp/dot-cargo
+          key: ${{ runner.os }}-${{ steps.git_info.outputs.commit_deps }}-dot-cargo
+
+      - run: git checkout ${{ steps.git_info.outputs.commit_deps }}
+
+      - name: Cargo cache (recovery)
+        if: steps.cache-dot-cargo.outputs.cache-hit == 'true'
+        run: |
+          rsync -av "/tmp/dot-cargo/" "$HOME/.cargo/"
+          find "$HOME/.cargo/" -exec touch -t "$(date +%Y%m%d%H%M.%S)" {} +
+
+      - name: Cargo cache (build)
+        run: cargo fetch --locked
+
+      - name: Cargo cache (save to cache)
+        if: steps.cache-dot-cargo.outputs.cache-hit != 'true'
+        run: rsync -av "$HOME/.cargo/" "/tmp/dot-cargo/"
+
+      # Caching external dependencies
+
+      - uses: actions/cache@v3
+        id: cache-deps
+        with:
+          path: /tmp/target-deps
+          key: ${{ runner.os }}-${{ steps.git_info.outputs.commit_deps }}-deps
+
+      - name: Dependencies (recovery)
+        if: steps.cache-deps.outputs.cache-hit == 'true'
+        run: |
+          rsync -av "/tmp/target-deps/" "$PWD/target/"
+          mtime=$(date +%Y%m%d%H%M.%S)
+          find ./target -exec touch -t "$mtime" {} +
+          find "/tmp/target-deps/" -exec touch -t "$mtime" {} +
+
+      - name: Dependencies (build)
+        run: cargo build --tests --release --frozen
+
+      - name: Dependencies (save to cache)
+        if: steps.cache-deps.outputs.cache-hit != 'true'
+        run: rsync -av "$PWD/target/" "/tmp/target-deps/"
+
+      # Caching canister code and tests
+
+      - uses: actions/cache@v3
+        id: cache-canister
+        with:
+          path: /tmp/target-canister
+          key: ${{ runner.os }}-${{ steps.git_info.outputs.commit_canister }}-canister
+
+      - run: git checkout ${{ steps.git_info.outputs.commit_canister }}
+
+      - name: Canister code (recover)
+        if: steps.cache-canister.outputs.cache-hit == 'true'
+        run: |
+          # Set timestamps so that --compare-dest below knows what (not) to save
+          find /tmp/target-canister -exec touch -t "$(date +%Y%m%d%H%M.%S)" {} +
+          rsync -av /tmp/target-canister/ "$PWD/target/"
+
+      - name: Canister code (build)
+        run: cargo build --tests --release --frozen
+
+      - name: Canister code (save to cache)
+        if: steps.cache-canister.outputs.cache-hit != 'true'
+        run: rsync -av --compare-dest="/tmp/target-deps/" "$PWD/target/" "/tmp/target-canister/"
+
+      # Build the tests
+      - name: Build
+        run: |
+          git checkout ${{ steps.git_info.outputs.commit_now }}
+          cargo build --tests --release --frozen
```

Comment:
Why do we need to run this multiple times? I would assumed, that if we restored all the files correctly (including timestamps), we could just run `cargo test` at the end?

Reply:
We could; though just like we have currently it's clearer to split build and tests so that we can quickly see what takes time